Cython/Compiler/ExprNodes.py

```diff
@@ -10446,13 +10502,14 @@ def analyse_types(self, env):
             return self.operand.coerce_to_boolean(env)
         to_py = self.type.is_pyobject
         from_py = self.operand.type.is_pyobject
+
         if from_py and not to_py and self.operand.is_ephemeral():
             if not self.type.is_numeric and not self.type.is_cpp_class:
                 error(self.pos, "Casting temporary Python object to non-numeric non-Python type")
         if to_py and not from_py:
             if self.type is bytes_type and self.operand.type.is_int:
                 return CoerceIntToBytesNode(self.operand, env)
-            elif self.operand.type.can_coerce_to_pyobject(env):
+            elif self.operand.type.can_coerce_to_pyobject(env) or isinstance(self.operand, SliceIndexNode):
```

Comment:
This condition looks fishy. Can we really always do the coercion whenever we find a `SliceIndexNode`? Does it no longer depend on the operand type in that case?

Reply:
Additional condition has been added